core/sql/src/main/java/org/trafodion/sql/HTableClient.java

```diff
@@ -421,29 +839,140 @@ else if (versions > 0)
+		    	  }
+		      }	    	
+		    }
+		    if (columnsToRemove != null)
+		    { //we are almost done checking if Exclude version of SingleColumnnValueFilter can be used. Th elast check s about to know if there is a IS_NULL_NULL
+		      //operation that cannot be using the Exclude method, as it is transformed in a filterList with OR, therefore we cannot guaranty that the SingleColumnValueExcludeFilter
+		      //performing the exclusion will be reached.
+		    	boolean is_null_nullFound = false;
+		    	for (Object o:compareOpList ){
+		    		if (new String((byte[])o).equals("IS_NULL_NULL")){
+		    			is_null_nullFound = true;
+		    			break;
+		    		}		    			
+		    	}
+		    	if (is_null_nullFound){
+		    		columnsToRemove = null; // disable Exclude method version of SingleColumnnValueFilter
+		    	}else
+		    		narrowDownResultColumns = false; // we will use the Exclude version of SingleColumnnValueFilter, so bypass the Family/QualifierFilter method
+		    }
+		    Filter f =constructV2Filter(colNamesToFilter,compareOpList,colValuesToCompare, columnsToRemove);
+		    if (f==null) return false; // error logging done inside constructV2Filter
+		    list.addFilter(f);
+		}//end V2
+		else{// deal with V1
+		    list = new FilterList(FilterList.Operator.MUST_PASS_ALL);
+		    
+		    for (int i = 0; i < colNamesToFilter.length; i++) {
+		      byte[] colName = (byte[])colNamesToFilter[i];
+		      byte[] coByte = (byte[])compareOpList[i];
+		      byte[] colVal = (byte[])colValuesToCompare[i];
+	
+		      if ((coByte == null) || (colVal == null)) {
+		        return false;
+		      }
+		      String coStr = new String(coByte);
+		      CompareOp co = CompareOp.valueOf(coStr);
+	
+		      SingleColumnValueFilter filter1 = 
+		          new SingleColumnValueFilter(getFamily(colName), getName(colName), 
+		              co, colVal);
+		      list.addFilter(filter1);
+		    }			
+		}//end V1
+	    // if we added a column for predicate eval, we need to filter down result columns
+	    FilterList resultColumnsOnlyFilter = null;
+	    if (narrowDownResultColumns){  	    	
+  		    HashMap<String,ArrayList<byte[]>> hm = new HashMap<String,ArrayList<byte[]>>(3);//use to deal with multiple family table
+  		    // initialize hm with list of columns requested for output
+  		    	for (int i=0; i<columns.length; i++){ // if we are here we know columns is not null
+  		    		if (hm.containsKey(new String(getFamily((byte[])columns[i])))){
+  		    			hm.get(new String(getFamily((byte[])columns[i]))).add((byte[])columns[i]);
+  		    		}else{
+  		    			ArrayList<byte[]> al = new ArrayList<byte[]>();
+  		    			al.add((byte[])columns[i]);
+  		    			hm.put(new String(getFamily((byte[])columns[i])), al);
+  		    		}	    			
+  		    	}
+  		    	
+  	    	if (hm.size()==1){//only one column family
+  	    		resultColumnsOnlyFilter = new FilterList(FilterList.Operator.MUST_PASS_ALL);
```

Comment:
With V2, you will see that the code that is used to compute columns is now happening after the code that is computing the predicate. Therefore the column list only contains the column needed in the returned result set. Where in V1 or no pushdown, the columns contain both the columns needed by the returned result set and the one needed for predicate evaluation.  There is no need to have a second list to store the one needed for predicate evaluation, as this information is inherently available in the colNameToFilter. That is why you see that with V2, I add the columns with the colNameToFilter to pass to the scan object, while with V1, colname is passed as is. For the resulting filter, well I am not sure if n**2 means n*2, where n is the columns in the result set only. But yes that is the price to pay to not get the whole thing, and I was indeed concerned about this, that is why I also implemented the SingleColumnValueExcludeFilter alternative whenever possible (this one is not adding any extra filter to do the job). Knowing when the filter is counter productive vs not could be part of a JIRA with the theme of how compiler can decide when push down is good vs counter productive. I intuit that given the poor efficiency of interpreted filter on the java side, vs high efficiency of native code on ESP side, for low selectivity predicate, pushdown will be counter productive... except may be for the column optimization, that should only significantly shine when huge amount of data have to make it to ESP... Anyway, given pred pushdown is disabled by default for now, it is imho OK to release with this, and condition enablement when we have compiler smartness to know when to turn it on or off or partially on (meaning don't try to optimize column)...